node/core/dispute-coordinator/src/initialized.rs

```diff
@@ -1186,6 +1191,89 @@ impl Initialized {
 
 		Ok(())
 	}
+
+	/// Firstly decrements spam slots for validators who voted on potential spam
+	/// candidates that are newly backed or included, and therefore no longer
+	/// potential spam.
+	/// 
+	/// Secondly queues participation if needed for freshly backed or included 
+	/// candidates. Backed candidates are added to best_effort queue while 
+	/// included are added to priority queue.
+	async fn handle_backed_included_events<Context>(
+		&mut self, 
+		ctx: &mut Context, 
+		overlay_db: &mut OverlayedBackend<'_, impl Backend>, 
+		now: u64,
+		candidate_events: Vec<CandidateEvent>,
+	)
+	{
+		let session = self.rolling_session_window.latest_session();
+		// Populate events contents, which allow us to queue participation without
+		// duplicated code
+		let mut events_contents: Vec<(CandidateReceipt, ParticipationPriority)> = Vec::new();
+		for event in candidate_events {
+			match event {
+				CandidateEvent::CandidateBacked(receipt, _, _, _) => {
+					events_contents.push((receipt, ParticipationPriority::BestEffort));
+				},
+				CandidateEvent::CandidateIncluded(receipt, _, _, _) => {
+					events_contents.push((receipt, ParticipationPriority::Priority));
+				}
+				_ => (),
+			}
+		}
+
+		for (receipt, priority) in events_contents {
+			// Clear spam slots
+			self.spam_slots.clear(&(session, receipt.hash()));
+
+			// Queue participation for freshly backed candidate. If dispute
+			// is already in priority queue don't try to queue participation
+			// in best_effort. Don't try to participate if candidate receipt
+			// can't be recovered from vote state, if we already voted, or
+			// if the candidate is not disputed.
+			let env =
+				match CandidateEnvironment::new(&*self.keystore, &self.rolling_session_window, session)
+			{
+				None => {
+					gum::warn!(
+						target: LOG_TARGET,
+						session,
+						"We are lacking a `SessionInfo` for handling import of statements."
+					);
+
+					continue // We skip queueing participation if not possible
+				},
+				Some(env) => env,
+			};
+			let votes_result = overlay_db.load_candidate_votes(session, &receipt.hash());
```

Comment:
Loading candidate votes from db on each backed and included candidate is costly unfortunately.

We should likely rather keep recent disputes in memory and then just do a quick lookup whether there is a dispute for that candidate ongoing - if no, then short circuit.

Reply:
On the other hand, we already do it for any backed candidate. So this should be acceptable for this PR. Still a followup where we only do this in case a candidate is actually disputed would be nice - we can and should have a call for this then.